Add neural network

0xsuu opened the pull request.

0xsuu committed:
Transfer prev work

Add NerualNetwork and an example.

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Merge remote-tracking branch 'origin/master' into feature/add-neural-network

0xsuu committed:
Add NN files

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
NN forward working

Signed-off-by: al1ensuu <462993145@qq.com>

review-ninja commented:
<a href="https://app.review.ninja/Llama-Group/Project-Llama/pull/138" target="_blank"><img src="https://app.review.ninja/63898156/pull/138/badge" alt="ReviewNinja"/>

0xsuu committed:
Done training, but still have bugs

Not train correctly.

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Fix most of them

Still need to debug to make the training work.

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Saving work

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Fix linux build error

Also, use clang++ as default compiler(for llama maybe), just a simple fix on linux.

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Save work for later

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Finished model, debugging

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Added function to calc total error

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Merge remote-tracking branch 'origin/master' into feature/add-neural-network

Signed-off-by: al1ensuu <462993145@qq.com>

Conflicts:
	llama_cpp/CMakeLists.txt

BlueCocoa committed:
Fix some typo and attribute((noreturn))

0xsuu committed:
Merge remote-tracking branch 'origin/fix/attribute-and-typo-for-logger' into feature/add-neural-network

0xsuu committed:
Change a little thing in formula

change the minus to plus.

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Linear neural network finished

Yay! Including some minor changes as well.

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Correctly working!

Works on and xor etc. All the logci operation.

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu committed:
Finish da shit off

Fixed a theory bug, now work perfectly!

Signed-off-by: al1ensuu <462993145@qq.com>

0xsuu commented:
LFGTM @BlueCocoa

0xsuu commented:
LGTM

BlueCocoa commented:
LGTM
